tasks/tasks/causemos_processors_test.py

```diff
@@ -0,0 +1,415 @@
+        "linear2": colinear2_values,
+        "constant": constant_values
+    })
+
+    return df
+
+
+
+roundToFour = partial(round, ndigits=4)
+
+
+def round_list(my_list):
+    return list(map(roundToFour, my_list))
+
+
+@pytest.mark.integration
+def test_perform_pca():
+
+    df = gen_control_dataframe(1000)
+
+    (pca_components_list, pca_components_count, pca_explained_ratios) = perform_pca(df)
+
+    assert pca_components_count == 5
+    assert list(map(lambda i: round(i, 4), pca_explained_ratios)) == [0.5004, 0.2602, 0.2393, 0.0, 0.0]
+
+    formatted = list(map(lambda i: round_list(i), pca_components_list.tolist()))
+
+    assert formatted == [
+        [0.0011, 0.0401, 0.7065, 0.7065, 0.0],
+        [0.7139, 0.6997, 0.0193, 0.0193, 0.0],
+        [0.7002, 0.7133, 0.0208, 0.0208, 0.0],
+        [0.0, 0.0, 0.7071, 0.7071, 0.0],
+        [0.0, 0.0, 0.0, 0.0, 1.0]
+    ]
+
+
+@pytest.mark.integration
+def test_pca_to_weights_v3_70_control_data():
+
+    pca_details = (
+        numpy.array(
+            [[1.13900827e-03, 4.01488796e-02, 7.06536188e-01,
+              7.06536188e-01, 0.00000000e+00],
+             [7.13923827e-01, 6.99690962e-01, 1.93044911e-02,
+              1.93044911e-02, 0.00000000e+00],
+             [7.00222444e-01, 7.13316638e-01, 2.08315041e-02,
+              2.08315041e-02, 0.00000000e+00],
+             [2.11047685e-17, 2.19081849e-17, 7.07106781e-01,
+              7.07106781e-01, 0.00000000e+00],
+             [0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
+              0.00000000e+00, 1.00000000e+00]]
+        ),
+
+        5,
+
+        numpy.array(
+            [5.00405570e-01, 2.60246340e-01, 2.39348091e-01,
+             2.51310581e-32, 0.00000000e+00]
+        )
+    )
+
+    weights = pca_to_weights_v3(pca_details)
+
+    rounded = list(map(lambda i: roundToFour(i), weights.tolist()))
+
+    assert rounded == [0.1685, 0.1829, 0.3243, 0.3243, 0.0]
+
+    assert sum(rounded) == 1
+
+
+
+@pytest.mark.integration
+def test_pca_to_weights_v3_90_control_data():
+
+    pca_details = (
+        numpy.array(
+            [[1.13900827e-03, 4.01488796e-02, 7.06536188e-01,
+              7.06536188e-01, 0.00000000e+00],
+             [7.13923827e-01, 6.99690962e-01, 1.93044911e-02,
+              1.93044911e-02, 0.00000000e+00],
+             [7.00222444e-01, 7.13316638e-01, 2.08315041e-02,
+              2.08315041e-02, 0.00000000e+00],
+             [2.11047685e-17, 2.19081849e-17, 7.07106781e-01,
+              7.07106781e-01, 0.00000000e+00],
+             [0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
+              0.00000000e+00, 1.00000000e+00]]
+        ),
+
+        5,
+
+        numpy.array(
+            [5.00405570e-01, 2.60246340e-01, 2.39348091e-01,
+             2.51310581e-32, 0.00000000e+00]
+        )
+    )
+
+    weights = pca_to_weights_v3(pca_details, 0.9)
+
+    rounded = list(map(lambda i: roundToFour(i), weights.tolist()))
+
+    assert rounded == [0.2434, 0.2565, 0.25, 0.25, 0.0]
```

Comment:
@brandomr this is the updated algorithm with the control data, with cumulative pca explained variance set to 90%.